pkg/envoy/registry/types.go

```diff
@@ -14,9 +15,13 @@ var log = logger.New("proxy-registry")
 type ProxyRegistry struct {
 	ProxyServiceMapper
 
-	connectedProxies sync.Map
+	connectedProxies map[string]*envoy.Proxy
```

Comment:
A raw map is not threadsafe. A consequence of this change is that the locks protecting this data structure are held longer. How confident are we that this change results in higher performance?

Reply:
I have similar doubt actually. I suggest doing a benchmark to compare these two approaches.

Reply:
there's no need to hold the lock longer, but the locks are locking on a single int, instead of on the (non-threadsafe) map itself.

It would be faster (no benchmark needed), simply grab the lock whenever reading/writing to the map

Reply:
My comment about holding the lock longer was referring to the PR's implementation since the lock was being acquired and then not released until the count was updated.

Are we sure this is a hot path though? How often are we checking the length of this map?

Reply:
I did a test comparing this the normal map v.s. original sync.Map. The test consists of two parts:

* Add 10000 proxies to the registry with 10000 goroutines
* When registry contains 10000 proxies, get the proxies count.

Both tests ensured the correctness of proxies count.

|  | sync.Map | normal map with mutex |
| --- | --- | --- |
| Add 10000 proxies | 9.5s | 0.26s |
| GetCount | 957642 ns | 13ns  |

Reply:
Great, thanks @allenlsy. That's what I was looking for